crates/ty_python_semantic/src/rank.rs

```diff
@@ -0,0 +1,78 @@
+//! A boxed bit slice that supports a constant-time `rank` operation.
+
+use bitvec::prelude::{BitBox, Msb0};
+
+/// A boxed bit slice that supports a constant-time `rank` operation.
+///
+/// This can be used to "shrink" a large vector, where you only need to keep certain elements, and
+/// you want to continue to use the index in the large vector to identify each element.
+///
+/// First you create a new smaller vector, keeping only the elements of the large vector that you
+/// care about. Now you need a way to translate an index into the large vector (which no longer
+/// exists) into the corresponding index into the smaller vector. To do that, you create a bit
+/// slice, containing a bit for every element of the original large vector. Each bit in the bit
+/// slice indicates whether that element of the large vector was kept in the smaller vector. And
+/// the `rank` of the bit gives us the index of the element in the smaller vector.
+///
+/// However, the naive implementation of `rank` is O(n) in the size of the bit slice. To address
+/// that, we use a standard trick: we divide the bit slice into 64-bit chunks, and when
+/// constructing the bit slice, precalculate the rank of the first bit in each chunk. Then, to
+/// calculate the rank of an arbitrary bit, we first grab the precalculated rank of the chunk that
+/// bit belongs to, and add the rank of the bit within its (fixed-sized) chunk.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct RankBitBox {
+    bits: BitBox<Chunk, Msb0>,
+    chunk_ranks: Box<[u32]>,
+}
+
+// bitvec does not support `u64` as a Store type on 32-bit platforms
+#[cfg(target_pointer_width = "64")]
+type Chunk = u64;
+#[cfg(not(target_pointer_width = "64"))]
+type Chunk = u32;
+
+const CHUNK_SIZE: usize = Chunk::BITS as usize;
+
+impl RankBitBox {
+    pub(crate) fn from_bits(iter: impl Iterator<Item = bool>) -> Self {
+        let bits: BitBox<Chunk, Msb0> = iter.collect();
+        let chunk_ranks = bits
+            .as_raw_slice()
+            .iter()
+            .scan(0u32, |rank, chunk| {
+                let result = *rank;
+                *rank += chunk.count_ones();
+                Some(result)
+            })
+            .collect();
+        Self { bits, chunk_ranks }
+    }
+
+    pub(crate) fn get_bit(&self, index: usize) -> Option<bool> {
+        self.bits.get(index).map(|bit| *bit)
+    }
+
+    /// Returns the number of bits _before_ (and not including) the given index that are set.
+    pub(crate) fn rank(&self, index: usize) -> u32 {
+        let chunk_index = index / CHUNK_SIZE;
```

Comment:
Hmm, too sad that `as_raw_slice` isn't const. I wonder if it is worth to mark this function and `get_bit` as  `#[inline]`